src/Microsoft.DotNet.Wpf/src/ReachFramework/Packaging/XpsFixedPageReaderWriter.cs

```diff
@@ -1866,26 +1866,24 @@ Uri         resourceUri
             Uri imageUri
             )
         {
-            //Extract file extension
+            //Extract file extension without '.'
             String path = imageUri.OriginalString;
-            String extension = Path.GetExtension(path).ToLower(CultureInfo.InvariantCulture);
-            //remove .
-            extension = extension.Substring(1);
+            ReadOnlySpan<char> extension = Path.GetExtension(path).ToLower(CultureInfo.InvariantCulture).AsSpan(1);
```

Comment:
Could you call `.ToLower(CultureInfo.InvariantCulture)` on the span instead of the string to avoid allocating a new string in ToLower.

Reply:
No, an in-place ToLower requires a mutable span.

Reply:
@stephentoub Is it faster than before when we do not use ToLower and check string Equals with `StringComparison.OrdinalIgnoreCase`?

Reply:
Sorry,  I don't understand the question.  Can you elaborate?

Reply:
@stephentoub  How about this code:

```
            //Extract file extension without '.'
            ReadOnlySpan<char> path = imageUri.OriginalString.AsSpan();
            ReadOnlySpan<char> extension = Path.GetExtension(path).Slice(1);
            ContentType contentType;
            if (extension.Equals(XpsS0Markup.JpgExtension, StringComparison.OrdinalIgnoreCase))
            {
                contentType = XpsS0Markup.JpgContentType;
            }

            // And so on

```

Is it faster?

Reply:
Ah, I understand the question now. The answer is, it depends.  It will save an allocation / copy, but it will need to do the casing operations for each comparison, and depending on how many of the comparisons it had to do, at some point that could become more expensive.  It would need to be measured, and if there's a cross-over point, weighed based on known use cases.

Reply:
@stephentoub Thank you. I wrote the test demo code and I found an improvement in performance.

|                                 Method |      Mean |    Error |    StdDev |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------------------------------- |----------:|---------:|----------:|-------:|------:|------:|----------:|
|           LookupContentTypeForImageUri | 114.17 ns | 3.096 ns |  9.130 ns | 0.0153 |     - |     - |      64 B |
| LookupContentTypeForImageUriIgnoreCase |  40.84 ns | 0.975 ns |  2.874 ns |      - |     - |     - |         - |
|           LookupContentTypeForImageUri | 115.00 ns | 3.858 ns | 11.375 ns | 0.0153 |     - |     - |      64 B |
| LookupContentTypeForImageUriIgnoreCase |  55.19 ns | 1.913 ns |  5.642 ns |      - |     - |     - |         - |
|           LookupContentTypeForImageUri |  84.97 ns | 1.773 ns |  3.198 ns | 0.0153 |     - |     - |      64 B |
| LookupContentTypeForImageUriIgnoreCase |  40.82 ns | 0.869 ns |  1.099 ns |      - |     - |     - |         - |
|           LookupContentTypeForImageUri |  91.29 ns | 1.693 ns |  1.322 ns | 0.0153 |     - |     - |      64 B |
| LookupContentTypeForImageUriIgnoreCase |  47.45 ns | 0.961 ns |  0.852 ns |      - |     - |     - |         - |
|           LookupContentTypeForImageUri | 103.35 ns | 1.876 ns |  2.234 ns | 0.0153 |     - |     - |      64 B |
| LookupContentTypeForImageUriIgnoreCase |  58.83 ns | 1.231 ns |  1.643 ns |      - |     - |     - |         - |
|           LookupContentTypeForImageUri |  52.88 ns | 0.973 ns |  0.911 ns | 0.0076 |     - |     - |      32 B |
| LookupContentTypeForImageUriIgnoreCase |  31.78 ns | 0.701 ns |  1.264 ns |      - |     - |     - |         - |
|           LookupContentTypeForImageUri |  65.51 ns | 1.325 ns |  1.361 ns | 0.0076 |     - |     - |      32 B |
| LookupContentTypeForImageUriIgnoreCase |  44.33 ns | 0.949 ns |  0.841 ns |      - |     - |     - |         - |
|           LookupContentTypeForImageUri |  66.49 ns | 1.383 ns |  2.386 ns | 0.0076 |     - |     - |      32 B |
| LookupContentTypeForImageUriIgnoreCase |  40.33 ns | 0.688 ns |  0.643 ns |      - |     - |     - |         - |
|           LookupContentTypeForImageUri |  76.48 ns | 1.560 ns |  1.669 ns | 0.0076 |     - |     - |      32 B |
| LookupContentTypeForImageUriIgnoreCase |  47.74 ns | 1.015 ns |  1.640 ns |      - |     - |     - |         - |
|           LookupContentTypeForImageUri |  83.05 ns | 1.536 ns |  1.769 ns | 0.0076 |     - |     - |      32 B |
| LookupContentTypeForImageUriIgnoreCase |  58.10 ns | 1.177 ns |  1.156 ns |      - |     - |     - |         - |


This is the benchmark code:


```csharp
    [MemoryDiagnoser]
    public class Program
    {
        static void Main(string[] args)
        {
            BenchmarkRunner.Run<Program>();
        }

        [Benchmark]
        [ArgumentsSource(nameof(GetArguments))]
        public object LookupContentTypeForImageUri(Uri imageUri)
        {
            //Extract file extension without '.'
            String path = imageUri.OriginalString;
            ReadOnlySpan<char> extension = Path.GetExtension(path).ToLower(CultureInfo.InvariantCulture).AsSpan(1);
            ContentType contentType;
            if (extension.Equals(XpsS0Markup.JpgExtension, StringComparison.Ordinal))
            {
                contentType = XpsS0Markup.JpgContentType;
            }
            else if (extension.Equals(XpsS0Markup.PngExtension, StringComparison.Ordinal))
            {
                contentType = XpsS0Markup.PngContentType;
            }
            else if (extension.Equals(XpsS0Markup.TifExtension, StringComparison.Ordinal))
            {
                contentType = XpsS0Markup.TifContentType;
            }
            else if (extension.Equals(XpsS0Markup.WdpExtension, StringComparison.Ordinal))
            {
                contentType = XpsS0Markup.WdpContentType;
            }
            else
            {
                //default to PNG
                contentType = XpsS0Markup.PngContentType;
            }
            return contentType;
        }

        [Benchmark]
        [ArgumentsSource(nameof(GetArguments))]
        public object LookupContentTypeForImageUriIgnoreCase(Uri imageUri)
        {
            //Extract file extension without '.'
            ReadOnlySpan<char> path = imageUri.OriginalString.AsSpan();
            ReadOnlySpan<char> extension = Path.GetExtension(path).Slice(1);
            ContentType contentType;
            if (extension.Equals(XpsS0Markup.JpgExtension, StringComparison.OrdinalIgnoreCase))
            {
                contentType = XpsS0Markup.JpgContentType;
            }
            else if (extension.Equals(XpsS0Markup.PngExtension, StringComparison.OrdinalIgnoreCase))
            {
                contentType = XpsS0Markup.PngContentType;
            }
            else if (extension.Equals(XpsS0Markup.TifExtension, StringComparison.OrdinalIgnoreCase))
            {
                contentType = XpsS0Markup.TifContentType;
            }
            else if (extension.Equals(XpsS0Markup.WdpExtension, StringComparison.OrdinalIgnoreCase))
            {
                contentType = XpsS0Markup.WdpContentType;
            }
            else
            {
                //default to PNG
                contentType = XpsS0Markup.PngContentType;
            }
            return contentType;
        }

        public IEnumerable<Uri> GetArguments()
        {
            yield return  new Uri(@"C:\foo.jpg") ;
            yield return  new Uri(@"C:\foo.JPG") ;
            yield return  new Uri(@"C:\foo.png") ;
            yield return  new Uri(@"C:\foo.PNG") ;
            yield return  new Uri(@"C:\foo.tif") ;
            yield return  new Uri(@"C:\foo.TIF") ;
            yield return  new Uri(@"C:\foo.wdp") ;
            yield return  new Uri(@"C:\foo.WDP") ;
            yield return  new Uri(@"C:\foo.none");
            yield return  new Uri(@"C:\foo.NONE");
        }
    }
```

You can find all the code in github

```
git init
git remote add origin https://github.com/lindexi/lindexi_gd.git
git pull origin 983c668ea7cd308a510c043b5d2e5de74da8efe7
cd ColalejayJearbearlerlekere
```